[updated] Incorporate Cajita / Cabana::Grid load balancer

cz4rs opened the pull request:
Updated version of  #104. I haven't touched the logic of it all, just a rebase, clean-up and dropping some redundant changes.

- resolved all conflicts and fixed deprecation warnings
- added `--sparse` command line flag for an unbalanced system to avoid having to change the source
- fixed data files naming scheme to facilitate processing the output with ParaView

aetx committed:
Add Cabana load balancing

cz4rs committed:
Fix deprecated code and remaining warnings

cz4rs committed:
Apply clang-format

cz4rs commented:
@streeve @aetx This is an updated version of PR #104 (see PR description for an overview of changes). The CI is pending approval, but based on the runs in my fork (https://github.com/cz4rs/CabanaMD/pull/1), all builds are passing.
Was there any reason that #104 was not merged?

For the context, we are exploring using CabanaMD as a benchmark for load balancing with communication.

aetx commented:
If I remember correctly, the only ugly hack was in `cabanamd_impl.h` to create an unbalanced load. By default, the whole system is uniformly filled (see https://github.com/ECP-copa/CabanaMD/pull/104#issuecomment-1030589568).

I don't remember another reason for it being not merged.

cz4rs committed:
Add a CLI flag for creating an unbalanced system

cz4rs committed:
Change file naming scheme

Place rank before timestep and add leading zeroes in the filenames.
This helps ParaView deduce the order correctly.

cz4rs committed:
Remove unused file

streeve commented:
Sorry I lost track of this PR and thanks very much for pushing this forward. Would you be okay with merging your general (and very helpful) changes in one PR, then merging the load balancing in a second?

cz4rs commented:
Sure, I will split the changes.
Please note that I'm on vacation now, so this will need to wait a bit (I'll
be back on September 9th).

pt., 30 sie 2024, 16:12 użytkownik Sam Reeve ***@***.***>
napisał:

> Sorry I lost track of this PR and thanks very much for pushing this
> forward. Would you be okay with merging your general (and very helpful)
> changes in one PR, then merging the load balancing in a second?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ECP-copa/CabanaMD/pull/113#issuecomment-2321392846>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABDMTPPUTPIEXHUC6CBM473ZUB4ULAVCNFSM6AAAAABJZQHOIWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGMRRGM4TEOBUGY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

cz4rs committed:
Use upstream Cabana repository

cz4rs committed:
Revert unnecessary changes

cz4rs commented:
@streeve Please have a look.
This is pretty much just a cleaned up version of #104. I have rebased this on top of `master` and cleaned up my commits a bit.

I have some more changes (extending LAMMPS commands support) in https://github.com/DARMA-tasking/CabanaMD/pull/3, I will post those in a separate PR.

cz4rs commented:
@streeve Ping

streeve commented on src/inputCL.cpp:
```diff
@@ -116,6 +116,8 @@ void InputCL::read_args( int argc, char *argv[] )
                  "                                (N = positive integer)\n",
                  "                                (PATH = location of ",
                  "directory)\n" );
+            log( std::cout, "  --sparse : "
```
I would prefer `--vacuum`, but not a big deal either way

cz4rs replied:
Renamed to `--vacuum`.

streeve commented on src/inputFile_impl.h:
```diff
@@ -455,7 +512,14 @@ void InputFile<t_System>::create_lattice( Comm<t_System> *comm )
     std::array<T_X_FLOAT, 3> global_low = { 0.0, 0.0, 0.0 };
     std::array<T_X_FLOAT, 3> global_high = { max_x, max_y, max_z };
-    system->create_domain( global_low, global_high );
+    if ( commandline.sparse )
```
Since this commandline arg is already set up, can we input the multiplier here instead of assuming 2?

cz4rs replied:
Sure, I have missed this somehow :man_facepalming:

streeve replied:
I think that was the original behavior to get results, but this will be much more general

cz4rs replied:
Done. Example below with `1.5` multiplier (`mpiexec -n 8 ./build/bin/cbnMD -il input/in.lj --vacuum 1.5`).

![image](https://github.com/user-attachments/assets/216fa763-a1b1-4ec4-a6f5-6156b471d43c)

streeve reviewed (commented):
Thanks for the reminder. Do you mind squashing @aetx's changes into a single "add Cabana load balancing" commit? Otherwise a couple very minor points and we can (finally) get this in

cz4rs committed:
Rework `--sparse` option

streeve reviewed (approved):
Looks good - thanks one more time to @aetx for the original work and @cz4rs for getting this over the finish line
